esp-bootloader-esp-idf/src/ota.rs

```diff
@@ -0,0 +1,463 @@
+//! # Over The Air Updates (OTA)
+//!
+//! ## Overview
+//! The OTA update mechanism allows a device to update itself based on data
+//! received while the normal firmware is running (for example, over Wi-Fi,
+//! Bluetooth or Ethernet).
+//!
+//! OTA requires configuring the Partition Tables of the device with at least
+//! two OTA app slot partitions (i.e., ota_0 and ota_1) and an OTA Data
+//! Partition.
+//!
+//! The OTA operation functions write a new app firmware image to whichever OTA
+//! app slot that is currently not selected for booting. Once the image is
+//! verified, the OTA Data partition is updated to specify that this image
+//! should be used for the next boot.
+//!
+//! Note: The prebuilt bootloaders provided by `espflash` _might not_ include
+//! OTA support. In that case you need to build the bootloader yourself.
+//!
+//! The general procedure to change the active slot
+//! - read the partition table [crate::partitions::read_partition_table]
+//! - find the Data/Ota partition
+//! - initialize [Ota]
+//! - read the current slot, change the current slot
+//!
+//! For more details see <https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/ota.html>
+use embedded_storage::{ReadStorage, Storage};
+
+use crate::partitions::FlashRegion;
+
+const SLOT0_DATA_OFFSET: u32 = 0x0000;
+const SLOT1_DATA_OFFSET: u32 = 0x1000;
+
+/// Representation of the current OTA slot.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::FromRepr)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Slot {
+    /// If there is a `firmware` app-partition it's used. Otherwise OTA-0
+    None,
+    /// OTA-0
+    Slot0,
+    /// OTA-1
+    Slot1,
+}
+
+impl Slot {
+    /// The slot represented as `usize`
+    pub fn number(&self) -> usize {
+        match self {
+            Slot::None => 0,
+            Slot::Slot0 => 0,
+            Slot::Slot1 => 1,
+        }
+    }
+
+    /// The next logical OTA slot
+    pub fn next(&self) -> Slot {
+        match self {
+            Slot::None => Slot::Slot0,
+            Slot::Slot0 => Slot::Slot1,
+            Slot::Slot1 => Slot::Slot0,
+        }
+    }
+
+    fn offset(&self) -> u32 {
+        match self {
+            Slot::None => SLOT0_DATA_OFFSET,
+            Slot::Slot0 => SLOT0_DATA_OFFSET,
+            Slot::Slot1 => SLOT1_DATA_OFFSET,
+        }
+    }
+}
+
+/// OTA image states for checking operability of the app.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Default, Hash, strum::FromRepr)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(u32)]
+pub enum OtaImageState {
+    /// Monitor the first boot. In bootloader this state is changed to
+    /// `PendingVerify`.
+    ///
+    /// You want to set this state after activating a newly installed update.
+    New           = 0x0,
+
+    /// First boot for this app was. This state is usually only set by the
```

Comment:
> First boot for this app was.

Yoda approves, but what does this mean? :) The docs on most of these variants could use a bit of revision.

Reply:
> The docs on most of these variants could use a bit of revision

Ok probably I can't hide where these are from anymore (and these are already revised)